kernel/queue.c

```diff
@@ -274,31 +274,39 @@ void k_queue_merge_slist(struct k_queue *queue, sys_slist_t *list)
 static void *k_queue_poll(struct k_queue *queue, s32_t timeout)
 {
 	struct k_poll_event event;
-	int err;
+	int err, elapsed = 0, done = 0;
 	unsigned int key;
 	void *val;
+	u32_t start;
 
 	k_poll_event_init(&event, K_POLL_TYPE_FIFO_DATA_AVAILABLE,
 			  K_POLL_MODE_NOTIFY_ONLY, queue);
 
+	if (timeout != K_FOREVER) {
+		start = k_uptime_get_32();
+	}
+
 	do {
 		event.state = K_POLL_STATE_NOT_READY;
 
-		err = k_poll(&event, 1, timeout);
-		if (err) {
+		err = k_poll(&event, 1, timeout - elapsed);
+
+		if (err && err != -EAGAIN) {
 			return NULL;
 		}
 
-		__ASSERT_NO_MSG(event.state ==
-				K_POLL_STATE_FIFO_DATA_AVAILABLE);
-
 		/* sys_sflist_* aren't threadsafe, so must be always protected
 		 * by irq_lock.
 		 */
 		key = irq_lock();
 		val = z_queue_node_peek(sys_sflist_get(&queue->data_q), true);
 		irq_unlock(key);
-	} while (!val && timeout == K_FOREVER);
+
+		if (!val && timeout != K_FOREVER) {
+			elapsed = k_uptime_get_32() - start;
```

Comment:
I always get a little concerned when I see uses of k_uptime_get_32() compared to one another.  The max value of an unsigned int would roll over ever 49 days.  What happens if the elapsed calculation gets caught in the roll over? (I'm pretty done is set to !0 ending the loop.)  Granted, the chances of getting an EAGAIN happen during an k_uptime_get_32() roll over are extremely slim ... but I thought I would bring it up.

Reply:
You could use a k_uptime_get() with s64_t types to eliminate the roll over issue entirely.  But it comes with s64_t handling cost.

Reply:
I'm not sure I understand an issue discussed here. I don't see "uses of k_uptime_get_32() compared to one another", only deltas between k_uptime_get_32() values being operated on. Otherwise, as @andyross mentioned, all Zephyr kernel timeout are expressed as 32-bit signed numbers anyway.

Reply:
@pfalcon It's not that the timeout is expressed as 32-bit number, it's that we're using a stored value of k_uptime_get_32() and subtracting it from another call of k_uptime_get_32().  If you happen to hit 49 days uptime between these 2 calls you will get a very odd result.

As @andyross mentioned, it's a pretty small chance that this will ever be hit.  I was just mentioning it as food for thought.  Similar to you mentioning the use of bool vs. int :)